File: dist-dependencies/SQLiteBatchCore.java
package io.sqlc;

public class SQLiteBatchCore {
  public interface BatchData {
    int getEntryCount();
    void openEntry(int index);
    String getEntryStatement();
    int getEntryBindColumnCount();
    boolean isEntryBindColumnNumber(int column);
    boolean isEntryBindColumnString(int column);
    double getEntryBindColumnDouble(int column);
    String getEntryBindColumnString(int column);
  }

  public interface BatchResults {
    void startNewEntry();
    void entryPutFieldAsInteger(String name, int value);
    void entryPutFieldAsString(String name, String value);
    void putNewEntry();
    void startNewEntryWithRows();
    void entryPutColumnName(String name);
    void startEntryRow();
    void entryPutRowValueAsString(String value);
    void entryPutRowValueAsDouble(double value);
    void entryPutRowValueAsNull();
    void entryPutRow();
    void putNewEntryWithRows();
  }

  static public int openBatchConnection(String fullName, int flags) {
    return SCCoreGlue.scc_open_connection(fullName, flags);
  }

  public static void
  executeBatch(final int mydbc, BatchData batchData, BatchResults batchResults) {
    try {
      final int count = batchData.getEntryCount();

      for (int i=0; i<count; ++i) {
        int previousTotalChanges = SCCoreGlue.scc_get_total_changes(mydbc);

        batchData.openEntry(i);

        String statement = batchData.getEntryStatement();

        if (SCCoreGlue.scc_begin_statement(mydbc, statement) != 0) {
          batchResults.startNewEntry();
          batchResults.entryPutFieldAsInteger("status", 1); // SQLite ERROR 1
          batchResults.entryPutFieldAsString("message",
            SCCoreGlue.scc_get_last_error_message(mydbc));
          batchResults.putNewEntry();
        } else {
          final int bindCount = batchData.getEntryBindColumnCount();

          int bindResult = 0; // SQLite OK

          for (int j = 0; j < bindCount; ++j) {
            if (batchData.isEntryBindColumnNumber(j)) {
              bindResult =
                SCCoreGlue.scc_bind_double(mydbc, 1 + j,
                  batchData.getEntryBindColumnDouble(j));
            } else if (batchData.isEntryBindColumnString(j)) {
              final String text = batchData.getEntryBindColumnString(j);
              bindResult =
                SCCoreGlue.scc_bind_text_utf16_bytes(mydbc, 1 + j,
                  text, text.length() * 2);
            } else {
              bindResult =
                SCCoreGlue.scc_bind_null(mydbc, 1 + j);
            }
          }

          if (bindResult != 0) {
            batchResults.startNewEntry();
            batchResults.entryPutFieldAsInteger("status", 1); // SQLite ERROR 1
            batchResults.entryPutFieldAsString("message",
            SCCoreGlue.scc_get_last_error_message(mydbc));
            batchResults.putNewEntry();
            SCCoreGlue.scc_end_statement(mydbc);
            continue;
          }

          int stepResult = SCCoreGlue.scc_step(mydbc);

          if (stepResult == 100) {
            final int columnCount = SCCoreGlue.scc_get_column_count(mydbc);

            batchResults.startNewEntryWithRows();

            for (int j=0; j < columnCount; ++j) {
              batchResults.entryPutColumnName(SCCoreGlue.scc_get_column_name(mydbc, j));
            }

            do {
              batchResults.startEntryRow();

              for (int col=0; col < columnCount; ++col) {
                final int type = SCCoreGlue.scc_get_column_type(mydbc, col);

                if (type == SCCoreGlue.SCC_COLUMN_TYPE_INTEGER ||
                    type == SCCoreGlue.SCC_COLUMN_TYPE_FLOAT) {
                  batchResults.entryPutRowValueAsDouble(
                    SCCoreGlue.scc_get_column_double(mydbc, col));
                } else if (type == SCCoreGlue.SCC_COLUMN_TYPE_NULL) {
                  batchResults.entryPutRowValueAsNull();
                } else {
                  batchResults.entryPutRowValueAsString(
                    SCCoreGlue.scc_get_column_text(mydbc, col));
                }
              }

              batchResults.entryPutRow();

              stepResult = SCCoreGlue.scc_step(mydbc);
            } while (stepResult == 100);

            batchResults.putNewEntryWithRows();
            SCCoreGlue.scc_end_statement(mydbc);
          } else if (stepResult == 101) {
            int totalChanges = SCCoreGlue.scc_get_total_changes(mydbc);
            int rowsAffected = totalChanges - previousTotalChanges;

            batchResults.startNewEntry();
            // same order as iOS & macOS ("osx"):
            batchResults.entryPutFieldAsInteger("status", 0); // SQLite OK
            batchResults.entryPutFieldAsInteger("totalChanges", totalChanges);
            batchResults.entryPutFieldAsInteger("rowsAffected", rowsAffected);
            batchResults.entryPutFieldAsInteger("lastInsertRowId",
              SCCoreGlue.scc_get_last_insert_rowid(mydbc));
            batchResults.putNewEntry();
            SCCoreGlue.scc_end_statement(mydbc);
          } else {
            batchResults.startNewEntry();
            batchResults.entryPutFieldAsInteger("status", 1); // SQLite ERROR 1
            batchResults.entryPutFieldAsString("message",
              SCCoreGlue.scc_get_last_error_message(mydbc));
            batchResults.putNewEntry();
            SCCoreGlue.scc_end_statement(mydbc);
          }
        }
      }
    } catch(Exception e) {
      // NOT EXPECTED - internal error:
      throw new RuntimeException(e);
    }
  }

  static {
    System.loadLibrary("sqlc-connection-core-glue");
    SCCoreGlue.scc_init();
  }
}

File: dist-dependencies/SQLiteBatchCore.m
#import "SQLiteBatchCore.h"

#include "sqlite-connection-core.h"

@implementation SQLiteBatchCore

+ (void) initialize
{
  scc_init();
}

+ (int) openBatchConnection: (NSString *) fullName
                      flags: (int) flags
{
  return scc_open_connection([fullName cString], flags);
}

+ (NSArray *) executeBatch: (int) connection_id
                      data: (NSArray *) data
{
  NSMutableArray * results = [NSMutableArray arrayWithCapacity: 0];

  for (int i=0; i < [data count]; ++i) {
    int previousTotalChanges = scc_get_total_changes(connection_id);

    NSArray * entry = [data objectAtIndex:i];

    NSString * statement = [entry objectAtIndex: 0];

    NSArray * bind = [entry objectAtIndex: 1];

    int prepareResult =
      scc_begin_statement(connection_id, [statement cString]);

    if (prepareResult == 0) {
      for (int j=0; j < [bind count]; ++j) {
        NSObject * bindValue = [bind objectAtIndex: j];

        if (bindValue == nil) {
          prepareResult = scc_bind_null(connection_id, 1 + j);
        } else if ([bindValue isKindOfClass: [NSNumber class]]) {
          prepareResult = scc_bind_double(connection_id, 1 + j,
            [(NSNumber *)bindValue doubleValue]);
        } else if ([bindValue isKindOfClass: [NSString class]]) {
          NSString * bindText = (NSString *)bindValue;
          prepareResult = scc_bind_text_utf8_bytes(connection_id, 1 + j,
              [bindText cString],
	      [bindText lengthOfBytesUsingEncoding: NSUTF8StringEncoding]);
        } else {
          prepareResult = scc_bind_null(connection_id, 1 + j);
        }
      }
    }

    if (prepareResult != 0) {
      [results addObject: @{
        @"status": @1,
        @"message": [NSString stringWithUTF8String:
          scc_get_last_error_message(connection_id)]
      }];

      scc_end_statement(connection_id);

      continue;
    }

    int stepResult = scc_step(connection_id);

    if (stepResult == 100) {
      const int columnCount = scc_get_column_count(connection_id);

      NSMutableArray * columns = [NSMutableArray arrayWithCapacity: 0];

      for (int j = 0; j < columnCount; ++j) {
        NSString * columnNameAsString =
          [NSString stringWithUTF8String: scc_get_column_name(connection_id, j)];
        [columns addObject: columnNameAsString];
      }

      NSMutableArray * rows = [NSMutableArray arrayWithCapacity: 0];

      do {
        NSMutableArray * row = [NSMutableArray arrayWithCapacity: 0];

        for (int j = 0; j < columnCount; ++j) {
          const int columnType = scc_get_column_type(connection_id, j);

          if (columnType == SCC_COLUMN_TYPE_INTEGER ||
              columnType == SCC_COLUMN_TYPE_FLOAT) {
            NSNumber * columnNumberValue =
              [NSNumber numberWithDouble: scc_get_column_double(connection_id, j)];
            [row addObject: columnNumberValue];
          } else if (columnType == SCC_COLUMN_TYPE_NULL) {
            [row addObject: [NSNull null]];
          } else {
            NSString * columnStringValue =
              [NSString stringWithUTF8String: scc_get_column_text(connection_id, j)];
            [row addObject: columnStringValue];
          }
        }

        [rows addObject: row];

        stepResult = scc_step(connection_id);
      } while(stepResult == 100);

      [results addObject: @{@"status":@0, @"columns": columns, @"rows": rows}];
    } else if (stepResult == 101) {
      int totalChanges = scc_get_total_changes(connection_id);
      int rowsAffected = totalChanges - previousTotalChanges;

      [results addObject: @{
        @"status": @0,
        @"rowsAffected": [NSNumber numberWithInteger: rowsAffected],
        @"totalChanges": [NSNumber numberWithInteger: totalChanges],
        @"lastInsertRowId": [NSNumber numberWithInteger:
          scc_get_last_insert_rowid(connection_id)]
      }];
    } else {
      [results addObject: @{
        @"status": @1,
        @"message": [NSString stringWithUTF8String:
          scc_get_last_error_message(connection_id)]
      }];
    }

    scc_end_statement(connection_id);
  }

  return results;
}

@end
